fix: update jpi version 0.43.0, gradle version 6.0.1, buildPlugin() command, and table-to-div migration

yasokada-lc opened the pull request:
fix: update gradle version 6.0.1
fix: update jpi version 0.43.0
fix: update buildPlugin command

<!-- Please describe your pull request here. -->

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your master branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

<!--
Put an `x` into the [ ] to show you have filled the information
-->

yasokada-lc committed:
fix: update gradle version 6.0.1
fix: update jpi version 0.43.0
fix: update buildPlugin command

yasokada-lc committed:
add test code

yasokada-lc commented:
In some use cases, multiple scm is needed instead of pipline.
Appropriate maintenance is desired.

nneul commented:
@yasokada-lc - does your MR happen to fix the complete breakage of this plugin in 2.277.x jenkins instances? Just filed JENKINS-65644 - with 2.277.x - looks like you cannot add a new SCM at all - the web UI just does nothing. No javascript error.

nneul commented:
Just tried building it locally - unfortunately, looks like it does not help the issue with 2.277.x

yasokada-lc commented:
I running test browser chrome or fire fox
Jenkins 2.263.3

2.277.x ???

oh no.
<img width="1390" alt="screen01-plugin-management" src="https://user-images.githubusercontent.com/4208784/118993420-77d15900-b9c0-11eb-93f0-1653158da810.png">
<img width="1390" alt="screen02-buildjob-sample" src="https://user-images.githubusercontent.com/4208784/118993427-79028600-b9c0-11eb-83f8-e1ed7136f1fb.png">

yasokada-lc commented:
Jenkins 2.263.3 running
but Jenkins 2.277.4 is loops ininity setTimeout(async)

Jenkins 2.263.3
<img width="989" alt="screen03-jenkins-2 263 3" src="https://user-images.githubusercontent.com/4208784/119002859-33e25200-b9c8-11eb-92ed-956afadac8ed.png">

Jenkins 2.277.4
<img width="1049" alt="screen04-jenkins-2 277 4" src="https://user-images.githubusercontent.com/4208784/119002856-3349bb80-b9c8-11eb-8b3e-ffb147c80a12.png">

yasokada-lc commented:
I updated the hetero-list-scm file and it works.

<img width="1011" alt="screen05-jenkins-2 277 4-ok" src="https://user-images.githubusercontent.com/4208784/119008812-74909a00-b9cd-11eb-95c1-a1f936e59e72.png">
<img width="1109" alt="screen06-jenkins-2 277 4-version-footerk" src="https://user-images.githubusercontent.com/4208784/119008813-75c1c700-b9cd-11eb-95b1-4be9cd007998.png">

yasokada-lc committed:
fix: hetero-list-scm update table tag to div tag.

nneul commented:
That fixes it for my 2.277.4 instance with latest build!

nneul commented:
Now question is what it would take to get someone to actually merge this......

nneul commented:
I've requested it in https://issues.jenkins.io/browse/JENKINS-65644 - hopefully someone will look at it.

nneul commented:
I did a build myself to validate on my server. You can try it if you want:

https://drive.google.com/file/d/1EapOnTjS8qzrZz6EpTMaTQ31oI6M_7n5/view?usp=sharing

nneul commented:
Quickest way to build if you want to be safe:  

```
git clone git@github.com:jenkinsci/multiple-scms-plugin.git
cd multiple-scms-plugin
git checkout feature/update-jpi
docker run --rm -u root -v "$PWD":/home/gradle/project -w /home/gradle/project gradle gradle build
```

result will be in build/libs

timja commented:
> Quickest way to build if you want to be safe:
> 
> ```
> git clone git@github.com:yasokada-lc/multiple-scms-plugin.git
> cd multiple-scms-plugin
> git checkout feature/update-jpi
> docker run --rm -u root -v "$PWD":/home/gradle/project -w /home/gradle/project gradle gradle build
> ```
> 
> result will be in build/libs

Or use the CI build 😄 https://ci.jenkins.io/job/Plugins/job/multiple-scms-plugin/job/PR-26/3/artifact/build/libs/multiple-scms.hpi

jimklimov committed:
Update build.gradle

Markup migration from table to div (without adaptive code to use either depending on Jenkins core capability) requires Jenkins 2.264+ which covers this year of weekly and "almost two" LTS releases by now (one pending soon). This constraint should prevent installation of the updated (and then incompatible) plugin to users who did/could not update their Jenkins core versions yet.

timja commented on build.gradle:
```diff
@@ -12,12 +12,13 @@ description = 'Allows multiple SCM plugins to be used in a job.'
     shortName = 'multiple-scms'
     displayName = 'Jenkins Multiple SCMs plugin'
-    url = 'https://wiki.jenkins-ci.org/display/JENKINS/Multiple+SCMs+Plugin'
+    url = 'https://plugins.jenkins.io/multiple-scms/'
```
this should be the github repo, it's used for documentation
```suggestion
    url = 'https://github.com/jenkins/multiple-scms-plugin'
```

timja replied:
the wiki content should be exported and added to the repo first before changing this URL https://jenkins-wiki-exporter.jenkins.io/

jimklimov replied:
Thanks, I'll address the docs in another PR then :)

jimklimov replied:
#27

jimklimov committed:
Update build.gradle

Bump the core version for first LTS with table-to-div migration support
Un-bump the documentation link to pursue in another PR

timja commented on src/test/java/org/jenkinsci/plugins/multiplescms/BuilderTest.java:
```diff
@@ -0,0 +1,24 @@
+    final String name = "Bobby";
+
+    @Test
+    public void testConfigRoundtrip() throws Exception {
```
this doesn't appear to do anything?

jimklimov replied:
I believe this satisfies a minimal formality level to get the plugin built - that there *is* some test to run and report to inspect.

jimklimov replied:
That said, I would welcome actual tests appearing eventually... though don't hold my breath for them with the plugin's outlook gloom as people migrate from legacy (freestyle etc.) jobs where this plugin really helps, into pipelines that can just use many checkout steps.

jimklimov reviewed (approved):
Thanks! Modulo the minimal required version of Jenkins core, looked good!

jimklimov commented:
And just for context, if someone would need a version of this plugin that works for older Jenkins Cores (1.x, 2.x before 2.264 if they can't upgrade their server) it may be possible to revive the adaptive markup from PR #25 which uses tabs or divs depending on core ability. PRs welcome then :)

garysturgeon commented:
Thanks for doing this! Already downloaded and it is working again.

enachi commented:
Wow! Finally! Can't wait to try it tomorrow! Thank a lot, everyone!

nneul commented:
Agreed, thank you very much @yasokada-lc  for putting this fix together and @jimklimov  for getting it merged quickly!

jimklimov commented:
Thanks @yasokada-lc - fixed the few legacy jobs on our server too, which we had to not touch (just in case to not corrupt the config) since our Jenkins core upgrade ;)
